ext/hash/hash.c

```diff
@@ -993,11 +973,75 @@ PHP_FUNCTION(mhash_keygen_s2k)
 
 #endif
 
+/* ----------------------------------------------------------------------- */
+
+/* {{{ php_hashcontext_create */
+static zend_object* php_hashcontext_create(zend_class_entry *ce) {
+	php_hashcontext_object *objval = ecalloc(1,
+		sizeof(php_hashcontext_object) + zend_object_properties_size(ce));
+	zend_object *zobj = &(objval->std);
+
+	zend_object_std_init(zobj, ce);
+	zobj->handlers = &php_hashcontext_handlers;
+
+	return zobj;
+}
+/* }}} */
+
+/* {{{ php_hashcontext_dtor */
+static void php_hashcontext_dtor(zend_object *obj) {
+	php_hashcontext_object *hash = php_hashcontext_from_object(obj);
+
+	/* Just in case the algo has internally allocated resources */
+	if (hash->context) {
+		unsigned char *dummy = emalloc(hash->ops->digest_size);
+		hash->ops->hash_final(dummy, hash->context);
+		efree(dummy);
+		efree(hash->context);
+		hash->context = NULL;
+	}
+
+	if (hash->key) {
+		memset(hash->key, 0, hash->ops->block_size);
```

Comment:
Indeed, I'll pile that on as a separate diff to keep the changes separate though.